packages/@aws-cdk/aws-codebuild/lib/project.ts

```diff
@@ -663,17 +666,10 @@ export class Project extends ProjectBase {
       }
 
       const sourceJson = this.source._toSourceJSON();
-      if (typeof buildSpec === 'string') {
-        return {
-          ...sourceJson,
-          buildSpec // Filename to buildspec file
-        };
-      } else if (Object.keys(buildSpec).length > 0) {
-        // We have to pretty-print the buildspec, otherwise
-        // CodeBuild will not recognize it as an inline buildspec.
+      if (buildSpec) {
         return {
           ...sourceJson,
-          buildSpec: JSON.stringify(buildSpec, undefined, 2)
+          buildSpec: buildSpec.toBuildSpec()
         };
       } else if (this.source.type === SourceType.None) {
         throw new Error("If the Project's source is NoSource, you need to provide a buildSpec");
```

Comment:
There is an interesting edge case here. I think the actual behavior should be to error out also if no source has been provided, and the build spec has been given as `BuildSpec.fromFileName()`. I'm not sure what would be a good way to accomplish that - perhaps a `public abstract validate(source: BuildSource): string[]` method on the `BuildSpec` class?

It's not super crucial though.